python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -156,1253 +120,308 @@ class MakeBrighterFatterKernelTaskConfig(pexConfig.Config):
-        """Prepare exposure for cross-correlation calculation.
-
-        For each amp, crop by the border amount, specified by
-        config.nPixBorderXCorr, then rescale by the gain
-        and subtract the sigma-clipped mean.
-        If the level is 'DETECTOR' then this is done
-        to the whole image so that it can be cross-correlated, with a copy
-        being returned.
-        If the level is 'AMP' then this is done per-amplifier,
-        and a copy of each prepared amp-image returned.
+                scaledCorrList.append(scaled)
+                self.log.info("Amp: %s %d/%d  Final: %g  XcorrCheck: %f",
+                              ampName, xcorrNum, len(xCorrList), q[0][0], xcorrCheck)
 
-        Parameters:
-        -----------
-        exp : `lsst.afw.image.exposure.ExposureF`
-            The exposure to prepare
-        gains : `lsst.cp.pipe.makeBrighterFatterKernel.BrighterFatterGain`
-            The object holding the amplifier gains, essentially a
-            dictionary of the amplifier gain values, keyed by amplifier name
-        level : `str`
-            Either `AMP` or `DETECTOR`
+            if len(scaledCorrList) == 0:
+                self.log.warn("Amp: %s All inputs rejected for amp!", ampName)
+                bfk.meanXcorrs[ampName] = np.zeros(bfk.shape)
+                bfk.ampKernels[ampName] = np.zeros(bfk.shape)
+                bfk.valid[ampName] = False
+                continue
 
-        Returns:
-        --------
-        scaledMaskedIms : `dict` [`str`, `lsst.afw.image.maskedImage.MaskedImageF`]
-            Depending on level, this is either one item, or n_amp items,
-            keyed by detectorId or ampName
-
-        Notes:
-        ------
-        This function is controlled by the following config parameters:
-        nPixBorderXCorr : `int`
-            The number of border pixels to exclude
-        nSigmaClipXCorr : `float`
-            The number of sigma to be clipped to
-        """
-        assert(isinstance(exp, afwImage.ExposureF))
+            if self.config.useAmatrix:
+                # This is mildly wasteful
+                preKernel = np.pad(self._tileArray(np.array(inputPtc.aMatrix[ampName])), ((1, 1)))
+            elif self.config.correlationQuadraticFit:
+                # Use a quadratic fit to the correlations as a function of flux.
+                preKernel = self.quadraticCorrelations(scaledCorrList, fluxes, f"Amp: {ampName}")
+            else:
+                # Use a simple average of the measured correlations.
+                preKernel = self.averageCorrelations(scaledCorrList, f"Amp: {ampName}")
 
-        local_exp = exp.clone()  # we don't want to modify the image passed in
-        del exp  # ensure we don't make mistakes!
+            center = int((bfk.shape[0] - 1) / 2)
 
-        border = self.config.nPixBorderXCorr
-        sigma = self.config.nSigmaClipXCorr
+            if self.config.forceZeroSum:
```

Comment:
Strictly, the sum should always be zero, due to charge/area conservation, correct? But we still allow the user to whether enforce this or not ?

Reply:
This is another thing kept to ensure the conversion didn't drop anything.  Looking over execution logs from imsim indicate that the sum is in the `10^-7 - 10^6` range.

Reply:
Why such a vast range? 13 orders of magnitude? Or did you mean `10^-6`? Probably the latter...